### DISCUSSION TITLE
Explore Logs Container: Add infinite scrolling to automatically load new log lines

### DISCUSSION THREAD

**Discussion by svennergr** [grafana/grafana#56092] [Observability]
Kind of similar to https://github.com/grafana/grafana/discussions/56091, but this discussion is about adding infinite scrolling to the LogPanel itself.
You easily lose the context to your last logline with the current behavior of loading "older" logs once you reached the end.

https://user-images.githubusercontent.com/8092184/193272595-bcd2611a-6929-44c9-a542-aa8eece1f45e.mov

**Comment by 09jvilla** [comment]
Assume we were to switch to infinite scrolling - is there any use case that now becomes harder? I'm trying to understand if there was any reason this was built the way it was in the first place. Something around being able to share a URL that always shows the same log lines maybe?

**Reply by gabor** [reply]
> Assume we were to switch to infinite scrolling - is there any use case that now becomes harder? I'm trying to understand if there was any reason this was built the way it was in the first place. Something around being able to share a URL that always shows the same log lines maybe?

@09jvilla not sure, i tried to come up with some disadvantages:
- if we want to make a good infinite-scroll, when the user is scrolling down, getting close to the bottom, we will run an additional Loki query to get more logs. this means more Loki queries will be sent, in a way that the user cannot really control, compared to the current situation (where the user explicitly has to click a button to get more log-lines)
- if we keep showing more logs "at the bottom", we need to decide what we will do "at the top":
     - we can just keep the log-lines at the top, so if you keep scrolling down, you will get more and more log-lines. this can affect the performance of grafana (we must have some kind of hard-limit on the number of log-lines to make sure the browser does not become unresponsive)
     - we can drop log-lines from "at the top" while we add new log lines "at the bottom", therefore maintaining a fixed size "window" into your logs. but then you loose your original log-line too, so i assume this does not help with the original problem of loosing the context.
- in the current approach, it was considered important that the user is able to search "in browser". by that i mean, you see 500log-lines in grafana in your browser, you can press CTRL+F in your browser, and search for something and it will work. this feature might get lost if we go deeper into infinite scrolling, for technical reasons. but, maybe this "feature" is not that important anymore?
- infinite-scroll is a more complex implementation (i mean the programming effort) compared to the current "paging" concept


i just want to make it clear, that this does not mean that i dislike the infinite-scroll idea. i was just trying to come up with any potential disadvantages. 😄

**Reply by 09jvilla** [reply]
This is super helpful @gabor and exactly what I was looking for. (I'm also secretly in favor of infinite scrolling, but trying to argue the other side). 

> in the current approach, it was considered important that the user is able to search "in browser". by that i mean, you see 500log-lines in grafana in your browser, you can press CTRL+F in your browser, and search for something and it will work. this feature might get lost if we go deeper into infinite scrolling, for technical reasons. but, maybe this "feature" is not that important anymore?

This one is probably the one that seems the most concerning but would be good to know if we would actually lose this functionality. Also I wonder if our existing instrumentation could tell us if people were doing this? I feel like if we make it easy enough to jump to keywords within the logs explore experience natively we could survive without this (or with a limited version of this). 

> but then you loose your original log-line too, so i assume this does not help with the original problem of loosing the context.

I'd say I'm less worried about this. The problem of losing context (at least based on the interview with the specific user we talked to) was the jumpiness between having a page of logs and then requesting the next one. You lose the continuity between the bottom of page X and the next line on page X+1. In an infinite scroll fixed window case, yes, we would start to lose some log lines as we add new ones, but we'd lose the ones furthest away from where the user is looking, and so the least relevant to their context.

**Reply by svennergr** [reply]
@09jvilla @gabor  regarding the "reason this was built the way it was in the first place": 
This document is a good read https://docs.google.com/document/d/1vr4S5Hmc2XMdK5daPZvawgD9P063THsK8rGP__F0xmA/edit#heading=h.od3uhe2uj0yw 

E.g. 
> When the user requests more logs we can either show the complete next subset (pagination), or combine it with the previous results (windowing). Pagination is easier to do, but the user loses context from what they have previously seen. If we want a windowing approach, we cannot grow the window forever because of browser limits. Instead we have to define our windowing mechanics, e.g., loading more loads half the limit, so that the result set is 1.5 times the limit. We could also try removing the other end's half results, so that the bottom old half and new results sum up to the original limit again. This would have the benefit of having one number as the mental constant for the windowing mechanics. Navigation to the present could use the previously cached data and drop older results. 

Infinite scrolling can not just "add" the next n-lines, because we might reach a limit the browser can handle. There are solutions - as mentioned - for this, but nevertheless this needs some work than the pagination we currently use.

**Comment by svennergr** [comment]
Just adding for reference. This discussion is also talking about better pagination: https://github.com/grafana/grafana/discussions/55935

**Comment by MichalKoziorowski-TomTom** [comment]
I understant CTRL+F is nice, but instead of that, user can modify query to add text filter. Other other option would be to duplicate query and to not loose original context add text filter to duplicated query. 

Or maybe some nice UI feature to add "search" that duplicates query and adds additional text filter. In this way user would loose completely nothing and work would be done. I'm pretty sure something cool could be designed.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
